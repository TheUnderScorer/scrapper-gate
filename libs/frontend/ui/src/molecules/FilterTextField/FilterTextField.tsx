import React, { Ref, useCallback, useEffect, useState } from 'react';
import {
  IconButton,
  InputAdornment,
  TextField,
  TextFieldProps,
} from '@material-ui/core';
import { ClearSharp } from '@material-ui/icons';
import { usePrevious } from 'react-use';
import { stopPropagation } from '@scrapper-gate/frontend/common';

export interface FilterTextFieldProps<T>
  extends Pick<
    TextFieldProps,
    'variant' | 'placeholder' | 'size' | 'className' | 'id'
  > {
  items: T[];
  onItemsChange: (items: T[]) => unknown;
  filterKeys: Array<keyof T>;
  onSearchChange?: (search: string) => unknown;
  ref?: Ref<HTMLInputElement>;
}

export const FilterTextField = <T extends unknown>({
  filterKeys,
  onItemsChange,
  items,
  onSearchChange,
  variant,
  ref,
  ...props
}: FilterTextFieldProps<T>) => {
  const [search, setSearch] = useState('');
  const prevSearch = usePrevious(search);
  const clearSearch = useCallback(() => setSearch(''), []);

  const filterItems = useCallback(
    (search: string) => {
      if (!search) {
        return onItemsChange(items);
      }

      const filteredItems = items.filter((item) => {
        return filterKeys.some((key) => {
          if (!item[key]) {
            return false;
          }

          return ((item[key] as unknown) as string)
            .toString()
            .toLowerCase()
            .includes(search.toLowerCase());
        });
      });

      onItemsChange(filteredItems);
    },
    [filterKeys, items, onItemsChange]
  );

  useEffect(() => {
    if (search === prevSearch) {
      return;
    }

    filterItems(search);
  }, [search, filterItems, prevSearch]);

  return (
    <TextField
      className="filter-text-field"
      ref={ref}
      onClick={stopPropagation}
      onKeyUp={stopPropagation}
      onKeyDown={stopPropagation}
      value={search}
      onChange={(event) => {
        setSearch(event.target.value);

        onSearchChange?.(event.target.value);
      }}
      InputProps={{
        endAdornment: (
          <InputAdornment position="end">
            <IconButton size="small" onClick={clearSearch}>
              <ClearSharp />
            </IconButton>
          </InputAdornment>
        ),
      }}
      variant={variant}
      {...props}
    />
  );
};
