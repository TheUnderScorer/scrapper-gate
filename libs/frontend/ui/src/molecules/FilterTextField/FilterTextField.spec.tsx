import { range } from 'remeda';
import { useState } from 'react';
import { List, ListItem } from '@material-ui/core';
import { act, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FilterTextField } from '../FilterTextField/FilterTextField';

interface TestItem {
  title: string;
  id: string;
}

const items: TestItem[] = range(0, 9).map((_, index) => ({
  title: `Item ${index}`,
  id: `#item_${index}`,
}));

const Component = (props: { filterKeys: Array<keyof TestItem> }) => {
  const [filteredItems, setFilteredItems] = useState(items);

  return (
    <>
      <FilterTextField
        items={items}
        onItemsChange={setFilteredItems}
        filterKeys={props.filterKeys}
      />
      <List>
        {filteredItems.map((item) => (
          <ListItem className="test-item" key={item.id} id={item.id}>
            {item.title}
          </ListItem>
        ))}
      </List>
    </>
  );
};

describe('<FilterTextField />', () => {
  it('should render without crashing', () => {
    const cmp = render(<Component filterKeys={['title']} />);

    expect(cmp).toMatchSnapshot();
  });

  it.each([
    ['title', 'Item 1'],
    ['id', '#item_3'],
  ])('should filter items', async (filterKey, text) => {
    const cmp = render(
      <Component filterKeys={[filterKey as keyof TestItem]} />
    );

    const field = cmp.container.querySelector('.filter-text-field input');

    act(() => {
      userEvent.type(field, text);
    });

    const testItems = cmp.container.querySelectorAll('.test-item');

    expect(testItems).toHaveLength(1);
  });
});
