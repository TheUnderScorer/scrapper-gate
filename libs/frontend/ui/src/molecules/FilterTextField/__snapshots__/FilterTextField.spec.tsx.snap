// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterTextField /> should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNone MuiTextField-root filter-text-field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-plngxy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-hiddenLabel MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-qlovab-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label css-4jkopv-MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ClearSharpIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1o2jng6-MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="css-1br1jja"
              style="width: 0.01px;"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <ul
        class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
          id="#item_0"
        >
          Item 0
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
          id="#item_1"
        >
          Item 1
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
          id="#item_2"
        >
          Item 2
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
          id="#item_3"
        >
          Item 3
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
          id="#item_4"
        >
          Item 4
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
          id="#item_5"
        >
          Item 5
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
          id="#item_6"
        >
          Item 6
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
          id="#item_7"
        >
          Item 7
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
          id="#item_8"
        >
          Item 8
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiFormControl-root MuiFormControl-marginNone MuiTextField-root filter-text-field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-plngxy-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-hiddenLabel MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-qlovab-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label css-4jkopv-MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ClearSharpIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1o2jng6-MuiOutlinedInput-notchedOutline"
          style="padding-left: 8px;"
        >
          <legend
            class="css-1br1jja"
            style="width: 0.01px;"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <ul
      class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
        id="#item_0"
      >
        Item 0
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
        id="#item_1"
      >
        Item 1
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
        id="#item_2"
      >
        Item 2
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
        id="#item_3"
      >
        Item 3
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
        id="#item_4"
      >
        Item 4
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
        id="#item_5"
      >
        Item 5
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
        id="#item_6"
      >
        Item 6
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
        id="#item_7"
      >
        Item 7
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters test-item css-fh7lmh-MuiListItem-root"
        id="#item_8"
      >
        Item 8
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
